vshard/storage/init.lua

```diff
@@ -3334,202 +3490,471 @@ local function service_call(service_name, ...)
 end
 
 --------------------------------------------------------------------------------
--- Configuration
+-- Master management
 --------------------------------------------------------------------------------
 
-local function storage_cfg(cfg, this_replica_uuid, is_reload)
-    if this_replica_uuid == nil then
-        error('Usage: cfg(configuration, this_replica_uuid)')
+local function master_role_update()
+    if this_is_master() and M.is_configured then
+        if not M.collect_bucket_garbage_fiber then
+            M.collect_bucket_garbage_fiber =
+                util.reloadable_fiber_create('vshard.gc', M, 'gc_bucket_f')
+        end
+        if not M.recovery_fiber then
+            M.recovery_fiber =
+                util.reloadable_fiber_create('vshard.recovery', M, 'recovery_f')
+        end
+    else
+        if M.collect_bucket_garbage_fiber then
+            M.collect_bucket_garbage_fiber:cancel()
+            M.collect_bucket_garbage_fiber = nil
+            log.info("GC stopped")
+        end
+        if M.recovery_fiber ~= nil then
+            M.recovery_fiber:cancel()
+            M.recovery_fiber = nil
+            log.info('Recovery stopped')
+        end
     end
-    local is_first_cfg = not M.current_cfg
-    cfg = lcfg.check(cfg, M.current_cfg)
-    local vshard_cfg, box_cfg = lcfg.split(cfg)
-    if vshard_cfg.weights or vshard_cfg.zone then
-        error('Weights and zone are not allowed for storage configuration')
+end
+
+local function master_on_disable()
+    log.info("Stepping down from the master role")
```

Comment:
When we disable master through config, we `sync` our replicaset:

```Lua
box.cfg({read_only = true})
sync(cfgctx.new_cfg.sync_timeout)
```

When master is automatically disabled, then no sync happens. Is it ok?

Frankly speaking, I don't understand the motivation of such sync on manual master switch too, it increases the time, in which several rebalancers can exist.

Reply:
Yes, this is intentional. I totally agree with you that the sync is probably not needed in both cases. However I didn't want to change the old behaviour even in this small detail. In the new behaviour I dropped the sync on purpose.